### PR TITLE
Refactor `cudf::contains`

### PR DESCRIPTION
**Pull request by ttnghia**
This is just a simple refactor to `cudf::contains`:
 * Remove `cudf/structs/detail/contains.hpp` and its corrresponding source file `src/structs/search/contains.cu`, moving its (modified) implementation into `src/search/contains_nested.cu`.
 * Adopt `experimental::row::equality::two_table_comparator` for struct equality comparison.
 * Add `const` qualifier for the `operator()` functions.
 * Rename some variables, and reorganize code to make it cleaner.

No new feature is added in this PR, just modifying the existing functions and moving things around.

This PR is extracted from the bigger PR for easier review. The original PR is https://github.com/rapidsai/cudf/pull/10656 for fully supporting nested type in `cudf::contains`. As such, this blocks it.

### CODE REVIEW

File: cpp/src/search/contains.cu [hunk 1]
```diff
@@ -14,22 +14,22 @@
  * limitations under the License.
  */
 
+#include <hash/unordered_multiset.cuh>
```
Comment (davidwendt) [inline cpp/src/search/contains.cu:17]: Why was this moved here?

File: cpp/src/search/contains.cu [hunk 2]
```diff
@@ -14,22 +14,22 @@
...
 #include <cudf/detail/nvtx/ranges.hpp>
 #include <cudf/detail/search.hpp>
 #include <cudf/dictionary/detail/search.hpp>
 #include <cudf/dictionary/detail/update_keys.hpp>
-#include <cudf/scalar/scalar_device_view.cuh>
-#include <cudf/search.hpp>
-#include <cudf/structs/detail/contains.hpp>
-#include <cudf/table/experimental/row_operators.cuh>
+#include <cudf/lists/list_view.hpp>
+#include <cudf/lists/lists_column_view.hpp>
```
Comment (davidwendt) [inline cpp/src/search/contains.cu:26]: I don't see where `lists_column_view.hpp` is needed.
Actually it would good to look at these headers again.
For example, I don't see where the `row_operators.cuh` or the `table_device_view.cuh` is needed in this source file. And it looks like we should be including `column_device_view.cuh` instead.

File: cpp/src/search/contains_nested.cu [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        });
+
+    } else {
+      return thrust::find_if(
+        rmm::exec_policy(stream), begin, end, [d_comp] __device__(auto const idx) {
+          return d_comp(idx,
+                        static_cast<rhs_index_type>(0));  // compare haystack[idx] == needle[0].
+        });
+    }
+  }();
```
Comment (davidwendt) [inline cpp/src/search/contains_nested.cu:68]: This seems an unnecessary IILE since the `found_it` is not really needed. Why not just the following?
```
if( haystack.has_nulls()) {
   ...
   return end != thrust::find_if(...);
}
return end != thrust::find_if(...);
```

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/src/search/contains.cu (hunk 1 above)
Comment (ttnghia) [reply cpp/src/search/contains.cu:17]: This is moved up to before `#include <cudf/...` headers, as it is in the `src/` folder.

File: cpp/src/search/contains.cu (hunk 2 above)
Comment (ttnghia) [reply cpp/src/search/contains.cu]: Thanks. I forgot to clean up the headers. Now that's done.

**Comment by ttnghia** [comment]
Rerun tests.

File: cpp/src/search/contains.cu [hunk 4]
```diff
@@ -14,155 +14,141 @@
...
-      auto found_iter = thrust::find(rmm::exec_policy(stream),  //
-                                     d_haystack->begin<Type>(),
-                                     d_haystack->end<Type>(),
-                                     s->value(stream));
+      auto const begin = d_haystack->begin<DType>();
+      auto const end   = d_haystack->end<DType>();
+      auto const val   = s->value(stream);
 
-      return found_iter != d_haystack->end<Type>();
+      return thrust::find(rmm::exec_policy(stream), begin, end, val) != end;
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/search/contains.cu:66]: I thought about this, but unfortunately not possible as `[begin, end, val]` are of different types depending on `has_nulls()`. The lambda must return the same type.

File: cpp/src/search/search_ordered.cu [hunk 5]
```diff
@@ -69,9 +67,8 @@ std::unique_ptr<column> search_ordered(table_view const& haystack,
 
   auto const comparator = cudf::experimental::row::lexicographic::two_table_comparator(
     matched_haystack, matched_needles, column_order, null_precedence, stream);
-  auto const has_null_elements =
-    has_nested_nulls(matched_haystack) or has_nested_nulls(matched_needles);
-  auto const d_comparator = comparator.device_comparator(nullate::DYNAMIC{has_null_elements});
+  auto const has_nulls    = has_nested_nulls(matched_haystack) or has_nested_nulls(matched_needles);
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/src/search/search_ordered.cu:70]: Yes, it is not better. I just feel the old name is too long :smile:.

File: cpp/include/cudf/detail/search.hpp [hunk 6]
```diff
@@ -61,15 +56,28 @@ std::unique_ptr<column> upper_bound(
 bool contains(column_view const& haystack, scalar const& needle, rmm::cuda_stream_view stream);
 
 /**
- * @copydoc cudf::contains(column_view const&, column_view const&, rmm::mr::device_memory_resource*)
- *
+
```
[the remark this answers is not in the document]
Comment (ttnghia) [reply cpp/include/cudf/detail/search.hpp]: Thanks for catching it. Done.

File: cpp/src/search/contains.cu [hunk 7]
```diff
@@ -14,155 +14,141 @@
...
 template <>
 bool contains_scalar_dispatch::operator()<cudf::struct_view>(column_view const& haystack,
                                                              scalar const& needle,
-                                                             rmm::cuda_stream_view stream)
+                                                             rmm::cuda_stream_view stream) const
 {
-  return cudf::structs::detail::contains(structs_column_view{haystack}, needle, stream);
+  CUDF_EXPECTS(haystack.type() == needle.type(), "scalar and column types must match");
+  // Haystack and needle structure compatibility will be checked by the table comparator
+  // constructor during calling to `contains_nested_element`.
```
Comment (davidwendt) [inline cpp/src/search/contains.cu:86]:
```suggestion
  // constructor during call to `contains_nested_element`.
```

**Review by davidwendt — APPROVED** [review]

**Review by Ethyling — APPROVED** [review]

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

**Review by PointKernel — APPROVED** [review]
